File: solutions/Symmetric-Tree/Symmetric-Tree.cpp

// /**
//  * Definition for binary tree
//  * struct TreeNode {
//  *     int val;
//  *     TreeNode *left;
//  *     TreeNode *right;
//  *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
//  * };
//  */
// class Solution {
// public:
//     bool isSymmetric(TreeNode *root) {
//         if(!root)
//             return true;
//         stack<TreeNode*>st1;
//         stack<TreeNode*>st2;
//         st1.push(root->left);
//         st2.push(root->right);
//         while(!st1.empty() && !st2.empty())
//         {
//             TreeNode *l=st1.top();
//             st1.pop();
//             TreeNode *r=st2.top();
//             st2.pop();
//             if(!l && !r)
//                 continue;
//             else if(l && r)
//             {
//                 if(l->val==r->val)
//                 {
//                     st1.push(l->left);
//                     st1.push(l->right);
//                     st2.push(r->right);
//                     st2.push(r->left);
//                 }
//                 else
//                     return false;
//             }
//             else
//                 return false;
//         }
//         return true;
//     }
// };


// Review - 20190726
// O(n)
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        if (!root)
        {
            return true;
        }
        else
        {
            return helper(root->left, root->right);
        }
    }
private:
    bool helper(TreeNode* l, TreeNode* r) {
        if (!l && !r)
        {
            return true;
        }
        else if (l && r)
        {
            return l->val == r->val && helper(l->left, r->right) && helper(l->right, r->left);
        }
        else
        {
            return false;
        }
    }
};

File: solutions/Symmetric-Tree/Symmetric-Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isSymmetric = function (root) {
  if (root === null) {
    return true;
  }
  const st1 = [root.left];
  const st2 = [root.right];
  while (st1.length > 0 && st2.length > 0) {
    const l = st1.pop();
    const r = st2.pop();
    if (l === null && r === null) {
      continue;
    } else if (l !== null && r !== null && l.val === r.val) {
      st1.push(l.left);
      st1.push(l.right);
      st2.push(r.right);
      st2.push(r.left);
    } else {
      return false;
    }
  }
  return true;
};